torch/csrc/jit/serialization/inlined_callstack_serialization.cpp

```diff
@@ -0,0 +1,212 @@
+     */
+    elements.reserve(3);
+    elements.emplace_back(debug_handle);
+    int64_t source_range_tag{-1};
+    const SourceRange& sr = it.second.first;
+    auto sr_it = source_range_tags.find(sr);
+    if (sr_it != source_range_tags.end()) {
+      source_range_tag = sr_it->second;
+    }
+    elements.emplace_back(source_range_tag);
+    elements.emplace_back(css_.serialize(it.second.second, source_range_tags));
+    c10::IValue tuple = c10::ivalue::Tuple::create(elements);
+    ivalues.emplace_back(tuple);
+  }
+  std::vector<at::Tensor> table;
+  c10::IValue ivalue = c10::ivalue::Tuple::create(std::move(ivalues));
+  auto result = jit::pickle(ivalue, &table);
+  TORCH_CHECK(table.size() == 0, "Expected 0 tensors to be written");
+  return result;
+}
+
+InlinedCallStackPtr InlinedCallStackDeserializer::deserialize(
+    const c10::IValue& iv,
+    const ska::flat_hash_map<int64_t, SourceRange>& source_range_map,
+    const std::shared_ptr<CompilationUnit>& cu) {
+  if (iv.isNone()) {
+    return c10::intrusive_ptr<InlinedCallStack>();
+  }
+  auto tup = iv.toTuple();
+  auto it = cached_inlined_callstacks_.find(tup);
+  if (it != cached_inlined_callstacks_.end()) {
+    return it->second;
+  }
+
+  auto tup_elems = tup->elements();
+  TORCH_INTERNAL_ASSERT(tup_elems.size() == 3);
+  // {IValue(module_instance_info), source_range_tag, IValue(InlinedCalddStack)}
+  auto module_instance_info =
+      deserialize_module_instance_info(tup_elems[0], cu);
+  int64_t source_range_tag = tup_elems[1].toInt();
+  auto source_range_it = source_range_map.find(source_range_tag);
+  TORCH_CHECK(
+      source_range_it != source_range_map.end(),
+      "Source range tag must exist in deserialized source range map.");
+  auto source_range = source_range_it->second;
+  auto callee = deserialize(tup_elems[2], source_range_map, cu);
+  InlinedCallStackPtr cs_ptr;
+  if (callee) {
+    cs_ptr = c10::make_intrusive<InlinedCallStack>(
+        callee, nullptr, source_range, module_instance_info);
+  } else {
+    cs_ptr = c10::make_intrusive<InlinedCallStack>(
+        nullptr, source_range, module_instance_info);
+  }
+  cached_inlined_callstacks_[tup] = cs_ptr;
+  // Invoking move constructor
+  // It is not clear if copy-ellision can happen since
+  // cs_ptr is copied into map above.
+  // This is to help avoid ref count update
+  return cs_ptr;
+}
+
+c10::optional<ModuleInstanceInfo> InlinedCallStackDeserializer::
+    deserialize_module_instance_info(
+        const c10::IValue& iv,
+        const std::shared_ptr<CompilationUnit>& cu) {
+  if (iv.isNone()) {
+    return c10::nullopt;
+  }
+  auto tup_elems = iv.toTuple()->elements();
+  TORCH_CHECK(tup_elems.size() == 2);
+  std::string type_name = tup_elems[0].toString()->string();
+  std::string instance_name = tup_elems[1].toString()->string();
+  // type_name might be empty string ""
+  // In that case type_ptr should be just nullptr
+  auto type_ptr = cu->get_class(type_name);
```

Comment:
Regarding other comment I made about why the CU is passed in: I see, you do need the compilation unit in order to get information for deserialization.